src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -42,12 +49,18 @@ private class ConfigurationCacheItem
             IEnumerable<IFeatureFilterMetadata> featureFilters,
             IEnumerable<ISessionManager> sessionManagers,
             ILoggerFactory loggerFactory,
-            IOptions<FeatureManagementOptions> options)
+            IOptions<FeatureManagementOptions> options,
+            IOptions<TargetingEvaluationOptions> assignerOptions,
+            IConfiguration configuration = null,
```

Comment:
I haven't seen the default parameter pattern with DI. Did you find some doc/example of this in use?

Reply:
I'm having difficulty finding a default pattern with DI. Every example I find says: "Dont do it". Or their suggestion doesn't quite work. I'm looking into it for the ITelemetryPublisher. Curious what the right pattern to follow is.

Reply:
I've seen what Ross is saying now, I found a couple sites mentioning it but now that I'm looking more it seems like it's not recommended just because you have to handle the nulls correctly, which I feel like isn't too bad in this case. Another option was to create empty versions of classes that inherit those interfaces and setting them as the default value instead of `null`. On [the documentation](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection#constructor-injection-behavior) for dependency injection it says it's ok to use default parameters.

Reply:
My immediate thought is remove the nullable constructor parameters and instead make them optional init properties. Then when registering in DI, use factory pattern for DI.

```
services.Add<IFeatureManager>(sp =>
    new FeatureManager(
        sp.GetRequiredService<...>(),
        ...)
        {
            Configuration = sp.GetService<IConfiguration>(), // May or may not exist in DI
            TargetingContextAccessor = sp.GetService<ITargetingContextAccessor>()
        });
```

Reply:
The properties are still null if no service is found though, right? In that case, I don't see how it would behave differently from the current code.

Reply:
It's not about behavioral difference. It's about design. With my suggestion, constructor parameters still represent requirements.

Reply:
Init properties are only available from .net 9 or above. Or is there a different way to declare them?

Reply:
@rossgrambo

C# 9, not .NET 9. C# version is just the language. I imagine everywhere we're building should have access to c# 9 compiler.

Reply:
Sorry, I meant C#. I was mainly concerned with our package having LangVersion set to 8.

https://github.com/microsoft/FeatureManagement-Dotnet/blob/main/src/Microsoft.FeatureManagement/Microsoft.FeatureManagement.csproj#L18

Is this something we can safely update?

Reply:
Yes, that setting is only used at build time.

Reply:
I'm wrong. C# 9 is only supported by runtimes .NET 5 and greater.

Reply:
This reminds me. .NET 3.1 is out of support and we need to drop that target in our next major release, aka this one.

So in fact, we should be able to use C# 9.

Reply:
I can't seem to get the `init` to work without errors unless I remove netstandard2.0 as well. Doesn't seem like we want to remove that though.

Reply:
Does it compile if you add the IsExternalInit type mentioned in this article https://sergiopedri.medium.com/enabling-and-using-c-9-features-on-older-and-unsupported-runtimes-ce384d8debb

Reply:
See [here](https://github.com/dotnet/roslyn/issues/45510#issuecomment-725091019)

Reply:
Oh my fault, yes that last link worked.

Reply:
Since we can work around this for lower .NET versions, we could remove .NET 3.1 in a different PR.